cloudinit/socket.py

```diff
@@ -0,0 +1,166 @@
+            raise ValueError(f"Invalid stage name: {stage}")
+        self.stage = stage
+        return self
+
+    def __enter__(self):
+        """Wait until a message has been received on this stage's socket.
+
+        Once the message has been received, enter the context.
+        """
+        if os.isatty(sys.stdin.fileno()):
+            LOG.info(
+                "Stdin is a tty, so skipping stage synchronization protocol"
+            )
+            return
+        sd_notify(
+            "STATUS=Waiting on external services to "
+            f"complete before starting the {self.stage} stage."
+        )
+        start_time = time.monotonic()
+        # block until init system sends us data
+        # the first value returned contains a message from the init system
+        #     (should be "start")
+        # the second value contains the path to a unix socket on which to
+        #     reply, which is expected to be /path/to/{self.stage}-return.sock
+        sock = self.sockets[self.stage]
+        chunk, self.remote = sock.recvfrom(5)
+
+        if b"start" != chunk:
+            # The protocol expects to receive a command "start"
+            self.__exit__(None, None, None)
+            raise ValueError(f"Received invalid message: [{str(chunk)}]")
+        elif f"{DEFAULT_RUN_DIR}/share/{self.stage}-return.sock" != str(
+            self.remote
+        ):
+            # assert that the return path is in a directory with appropriate
+            # permissions
+            self.__exit__(None, None, None)
+            raise ValueError(f"Unexpected path to unix socket: {self.remote}")
+
+        total = time.monotonic() - start_time
+        time_msg = f"took {total: .3f}s to " if total > 0.01 else ""
+        sd_notify(f"STATUS=Running ({self.stage} stage)")
+        LOG.debug("sync(%s): synchronization %scomplete", self.stage, time_msg)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Notify the socket that this stage is complete."""
+        message = "done"
+        systemd_exit_code = "0"
+        if exc_type:
+            # handle exception thrown in context
+            systemd_exit_code = "1"
+            status = f"{repr(exc_val)} in {exc_tb.tb_frame}"
+            message = (
+                'fatal error, run "systemctl status cloud-init-main.service" '
+                'and "cloud-init status --long" for more details'
+            )
+            if not self.first_exception:
+                self.first_exception = message
+            LOG.fatal(status)
+            sd_notify(f"STATUS={status}")
+
+        sock = self.sockets[self.stage]
+        sock.connect(self.remote)
+
+        # the returned message will be executed in a subshell
+        # hardcode this message rather than sending a more informative message
+        # to avoid having to sanitize inputs (to prevent escaping the shell)
+        sock.sendall(f"echo '{message}'; exit {systemd_exit_code};".encode())
```

Comment:
New comment, same section here. This exit code appears to be getting swallowed up.
I've we raise and exception in any individual stage (like a ZeroDivisionError in your example). That exception is being swallowed by status_wrapper in cloudinit.cmd.main in order to write out result.json/status.json. So, we generally will find that, if our former stages call cloud-init CLI directly instead of socket communication, this ZeroDivisionError that affects both cloud-config.service and cloud-final.service would have resulted reporting two failed units from systemctl list-units --failed. If we allow the stages to interact with cloud-init-main's sockets, and we add `sh -ex` to debug the expected script that was produced by socket.py we can see that exit 0 is produced in all cases, and no exc_type exc_val or exc_tb are produced as we leave to SocketSync.context manager. I'm trying to provide a reasonable approach to treating exceptions when in singleprocess mode that would involve raising that exception.


Maybe we actually need to check the return codes of sub_main in each sync'd stage?
```diff
diff --git a/cloudinit/cmd/main.py b/cloudinit/cmd/main.py
index d5d2fa79d..2b8d49ec9 100644
--- a/cloudinit/cmd/main.py
+++ b/cloudinit/cmd/main.py
@@ -1190,7 +1190,9 @@ def all_stages(parser):
         args = parser.parse_args(args=["init", "--local"])
         args.skip_log_setup = False
         # run local stage
-        sub_main(args)
+        if not sub_main(args):
+            # Possibly read status.json[<stage>]['errors']?
+            raise RuntimeError("Something broke local stage")
 
     # wait for cloud-init-network.service to start
     with sync("network"):
@@ -1198,7 +1200,9 @@ def all_stages(parser):
         args = parser.parse_args(args=["init"])
         args.skip_log_setup = True
         # run init stage
-        sub_main(args)
+        if not sub_main(args):
+            # Possibly read status.json[<stage>]['errors']?
+            raise RuntimeError("Something broke init stage")
 
     # wait for cloud-config.service to start
     with sync("config"):
@@ -1206,7 +1210,9 @@ def all_stages(parser):
         args = parser.parse_args(args=["modules", "--mode=config"])
         args.skip_log_setup = True
         # run config stage
-        sub_main(args)
+        if not sub_main(args):
+            # Possibly read status.json[<stage>]['errors']?
+            raise RuntimeError("Something broke config stage")
 
     # wait for cloud-final.service to start
     with sync("final"):
@@ -1214,7 +1220,9 @@ def all_stages(parser):
         args = parser.parse_args(args=["modules", "--mode=final"])
         args.skip_log_setup = True
         # run final stage
-        sub_main(args)
+        if not sub_main(args):
+            # Possibly read status.json[<stage>]['errors']?
+            raise RuntimeError("Something broke final stage")
 
     # signal completion to cloud-init-main.service
     if sync.first_exception:
```

Reply:
> New comment, same section here. This exit code appears to be getting swallowed up. 

Nothing is swallowed up that wasn't before.

> I've we raise and exception in any individual stage (like a ZeroDivisionError in your example). That exception is being swallowed by status_wrapper in cloudinit.cmd.main in order to write out result.json/status.json.

That's correct. Cloud-init continues to treat exceptions as it did before, but with an added later of handling that is only used to handle exceptions that occur in the socket synchronization protocol.

Error messages in `systemctl status <stage>` when a stage experiences a failure _below_ the synchronization protocol aren't going to happen currently - and that wasn't the intent. The point of this exception handling is to deal with exceptions in the synchronization layer itself. We can always add those messages if we want them later into status_wrapper or whatever, but that's not the point of this PR.

> So, we generally will find that, if our former stages call cloud-init CLI directly instead of socket communication, this ZeroDivisionError that affects both cloud-config.service and cloud-final.service would have resulted reporting two failed units from systemctl list-units --failed.

Huh? How could cloud-config.service and cloud-final.service signal error?

The failed stage would signal failure - this is to signal which stage the boot failure happened during.

Additionally, the `cloud-init-main.service` will signal failure, because this is where the code that experienced the failure actually failed.

> If we allow the stages to interact with cloud-init-main's sockets, and we add `sh -ex` to debug the expected script that was produced by socket.py we can see that exit 0 is produced in all cases, and no exc_type exc_val or exc_tb are produced as we leave to SocketSync.context manager. I'm trying to provide a reasonable approach to treating exceptions when in singleprocess mode that would involve raising that exception.
> 
> Maybe we actually need to check the return codes of sub_main in each sync'd stage?

This would give up on any opportunity to run future stages. The point of this was to attempt to continue on as a best effort in degraded state to avoid preventing the system from continuing to boot in a degraded state.

Reply:
@blackboxsw let me know what you think of the latest few commits. This works by capturing the exit code from the stages beneath them and merging that into the one that already gets reported to systemd - as well as making sure that the main service exits non-zero if any substage exits non-zero.

Reply:
@holmanb this is excellent, thanks for perseverance here. I think this provides enough of a breadcrumb into failing the expected stage where we saw the error. Additional failure details and context can be seen from `cloud-init status --format=yaml ` to allow folks to debug the issue more deeply.
```
root@test-n:~# systemctl list-units --failed
  UNIT                 LOAD   ACTIVE SUB    DESCRIPTION             
● cloud-config.service loaded failed failed Cloud-init: Config Stage

Legend: LOAD   → Reflects whether the unit definition was properly loaded.
        ACTIVE → The high-level unit activation state, i.e. generalization of SUB.
        SUB    → The low-level unit activation state, values depend on unit type.

1 loaded units listed.

# systemctl status cloud-config
× cloud-config.service - Cloud-init: Config Stage
     Loaded: loaded (/usr/lib/systemd/system/cloud-config.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Wed 2024-07-31 18:22:25 UTC; 5min ago
    Process: 292 ExecStart=sh -c echo "start" | nc.openbsd -Uu -W1 /run/cloud-init/share/config.sock -s /run/cloud-init/share/config-return.sock | sh (code=exited, status=1/FAILURE)
   Main PID: 292 (code=exited, status=1/FAILURE)
        CPU: 10ms

Jul 31 18:22:24 test-n systemd[1]: Starting cloud-config.service - Cloud-init: Config Stage...
Jul 31 18:22:25 test-n sh[299]: done
Jul 31 18:22:25 test-n systemd[1]: cloud-config.service: Main process exited, code=exited, status=1/FAILURE
Jul 31 18:22:25 test-n systemd[1]: cloud-config.service: Failed with result 'exit-code'.
Jul 31 18:22:25 test-n systemd[1]: Failed to start cloud-config.service - Cloud-init: Config Stage.
```